File: cpp/08_027.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

int maxSubarrayWithTargetSum(
    const std::vector<int>& nums, 
    int target
) {
    // Map to store the first occurrence of each sum
    std::unordered_map<int, int> sumIndexMap;
    sumIndexMap[0] = -1; // Initialize with 0 at index -1
    
    int maxLength = 0;
    int currentSum = 0;
    
    for (int i = 0; i < nums.size(); i++) {
        currentSum += nums[i];
        
        // Check if we can form a subarray with sum = target
        if (sumIndexMap.find(currentSum - target) != 
            sumIndexMap.end()) {
            int subarrayStart = sumIndexMap[currentSum - target] + 1;
            int currentLength = i - subarrayStart + 1;
            maxLength = std::max(maxLength, currentLength);
        }
        
        // Only store the first occurrence of each sum
        if (sumIndexMap.find(currentSum) == 
            sumIndexMap.end()) {
            sumIndexMap[currentSum] = i;
        }
    }
    
    return maxLength;
}

int main() {
    std::vector<int> nums = {1, -1, 5, -2, 3, 0, 2, -4, 1};
    int target = 3;
    std::cout << maxSubarrayWithTargetSum(nums, target) 
              << std::endl; // Output: 4
    return 0;
}

File: java/08_027.java
import java.util.HashMap;
import java.util.Map;

/**
 * Finds maximum length subarray with a target sum
 */
public class MaximumSubarrayWithTargetSum {
   /**
    * Find length of longest subarray with sum equal to target
    */
   public static int maxSubarrayWithTargetSum(int[] nums, 
         int target) {
      // Track first occurrence of each cumulative sum
      Map<Integer, Integer> sumIndexMap = new HashMap<>();
      sumIndexMap.put(0, -1); // Sum 0 at position -1
      
      int maxLength = 0;
      int currentSum = 0;
      
      for (int i = 0; i < nums.length; i++) {
         // Update running sum
         currentSum += nums[i];
         
         // Check if we can form target sum subarray
         if (sumIndexMap.containsKey(currentSum - target)) {
            int subarrayStart = sumIndexMap.get(currentSum - target) + 1;
            int currentLength = i - subarrayStart + 1;
            maxLength = Math.max(maxLength, currentLength);
         }
         
         // Only store first occurrence of each sum
         if (!sumIndexMap.containsKey(currentSum)) {
            sumIndexMap.put(currentSum, i);
         }
      }
      
      return maxLength;
   }
   
   /**
    * Test the algorithm
    */
   public static void main(String[] args) {
      int[] nums = {1, -1, 5, -2, 3, 0, 2, -4, 1};
      int target = 3;
      int result = maxSubarrayWithTargetSum(nums, target);
      System.out.println("Maximum subarray length: " + result);
   }
}